### PR TITLE
Removing an operation with mv command does not remove the operation

### PR DESCRIPTION
**Pull request by PradeepKiruvale**
## Proposed changes

Issue:
Removing an operation file using the `mv` command does not remove it from the tedge-mapper's operations list.
The reason for this is when a file is moved, a `file modified` event gets generated and this will be considered as an `add` operation. So, the mapper tries to access the file and fails to read the file and add it to the list, instead of removing the operation from the list.

Proposed solution:

- So, the proposed solution is when a `file is modified` and a `modified/Add` event is generated, and if the operation file does not exist then remove the operation from the list.- 
- Also replaced the operation list, with a map to store the operations. A single map is used to store the operation.- 
- Populate the operations map every time there is an event, irrespective of add/remove events.

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue

https://github.com/thin-edge/thin-edge.io/issues/1974#:~:text=Removing%20an%20operation%20with%20mv%20command%20does%20not%20remove%20the%20operation%20%231974

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Review by didier-wenzek — COMMENTED** [review]
> Removing an operation file using the mv command does not remove it from the list.
The reason for this is when a file is moved, then the file modified event gets generated and this will be considered as an add operation and tries to access the file and fails to read the file instead of removing the operation from the list.

I don't understand the issue. Do you mean that the `Remove` event is not sent to the mapper? If so, what needs to be fix is the |`inotify` support.

**Comment by PradeepKiruvale** [comment]
> > Removing an operation file using the mv command does not remove it from the list.
> > The reason for this is when a file is moved, then the file modified event gets generated and this will be considered as an add operation and tries to access the file and fails to read the file instead of removing the operation from the list.
> 
> I don't understand the issue. Do you mean that the `Remove` event is not sent to the mapper? If so, what needs to be fix is the |`inotify` support.

Yes, when a file is removed from the `/etc/tedge/operations/c8y` directory using the `mv` command, the `Remove/delete` event is not sent to the mapper, instead `modified` event is generated, and sent to the mapper.

File: crates/extensions/c8y_mapper_ext/src/converter.rs [hunk 1]
```diff
@@ -791,9 +791,17 @@ fn add_or_remove_operation(
...
             let ops_dir = message.ops_dir.clone();
             let op_name = message.operation_name.clone();
-            let op = get_operation(ops_dir.join(op_name))?;
-
-            ops.add_operation(op);
+            match get_operation(ops_dir.join(op_name)) {
+                Ok(op) => ops.add_operation(op),
+                Err(e) => {
+                    if e.to_string().eq("No such file or directory (os error 2)") {
+                        ops.remove_operation(&message.operation_name);
```
[the remark this answers is not in the document]
Comment (PradeepKiruvale) [reply crates/extensions/c8y_mapper_ext/src/converter.rs:798]: Agree, It is better to have an `HashMap` for operation.

**Comment by PradeepKiruvale** [comment]
> > > Removing an operation file using the mv command does not remove it from the list.
> > > The reason for this is when a file is moved, then the file modified event gets generated and this will be considered as an add operation and tries to access the file and fails to read the file instead of removing the operation from the list.
> > 
> > 
> > I don't understand the issue. Do you mean that the `Remove` event is not sent to the mapper? If so, what needs to be fix is the |`inotify` support.

Do you mean to say that we need to fix it in our `fs notify wrapper` library or the actual `notify` crate?

> 
> Yes, when a file is removed from the `/etc/tedge/operations/c8y` directory using the `mv` command, the `Remove/delete` event is not sent to the mapper, instead `modified` event is generated, and sent to the mapper.

**Comment by PradeepKiruvale** [comment]
> How about a simpler approach where we **don't** differentiate between `Add/Remove/Modified` events separately but instead always reload the contents of the operations dir whenever any update happens on that directory? This prevents the `operations_by_trigger` map from diverging from the reality on the file system. I don't think the cost of regenerating the map by reading the whole directory content instead of selectively adding or removing from the existing map is a big deal here as adding/removing operations is such a rare operation.

For me, it makes sense as there will not be many operations to be loaded and the change is not that frequent. @reubenmiller  @didier-wenzek  what do you guys think about Albin's suggestion?

File: crates/extensions/c8y_mapper_ext/src/converter.rs [hunk 2]
```diff
@@ -746,7 +746,9 @@ impl Converter for CumulocityConverter {
             .ops_dir
             .eq(&self.cfg_dir.join("operations").join("c8y"))
         {
-            add_or_remove_operation(message, &mut self.operations)?;
+            // Re populate the operations irrespective add/remove/modify event
+            self.operations = get_operations(message.ops_dir.clone())?;
+            dbg!(&self.operations);
```
Comment (albinsuresh) [inline crates/extensions/c8y_mapper_ext/src/converter.rs:751]: Don't forget to remove that before the final commit.

**Review by albinsuresh — APPROVED** [review]

**Review by didier-wenzek — CHANGES_REQUESTED** [review]
I'm okay to reload the whole operation set when there is a change. However, the `Operations` data structure must be revised. The current version is wrongly using a hashmap as a vector.

File: crates/core/c8y_api/src/smartrest/operations.rs [hunk 3]
```diff
@@ -64,75 +67,34 @@ pub fn is_valid_operation_name(operation: &str) -> bool {
...
+        let ops: Vec<Operation> = self.operations.clone().into_values().collect();
+        for op in ops {
+            if let Some(template) = op.template() {
+                if template.eq(operation_template) {
+                    return Some(op);
+                }
+            }
+        }
+        None
     }
```
[the remark this answers is not in the document]
Comment (PradeepKiruvale) [reply crates/core/c8y_api/src/smartrest/operations.rs]: Thats right. Replaced the map with a vector.

File: crates/extensions/c8y_mapper_ext/src/converter.rs (hunk 2 above)
Comment (didier-wenzek) [reply crates/extensions/c8y_mapper_ext/src/converter.rs:751]: Still to be removed ;-)

Comment (PradeepKiruvale) [reply crates/extensions/c8y_mapper_ext/src/converter.rs]: done

**Review by didier-wenzek — APPROVED** [review]
Approved

**Pull request merged**